### PR TITLE
Fix imports plt in Dalvik and other stuff

### PR DESCRIPTION
**Pull request by h4ng3r**
R2R PR: https://github.com/radare/radare2-regressions/pull/582

BUG INFO: 
- BT info: http://pastebin.com/AmvwyY4f
- fcn_recurse: http://pastebin.com/YGESnb8j

### CODE REVIEW

**Comment by h4ng3r** [comment]
I have a partial fix:
```r2 -qc"ii;is" radare2-regressions/bins/dex/HelloWorld.dex
[Imports]
ordinal=000 plt=0x000000b9 bind=NONE type=FUNC name=Ljava/io/PrintStream.method.println(Ljava/lang/String;)V
ordinal=000 plt=0x000000c1 bind=NONE type=FUNC name=Ljava/lang/Integer.method.<init>(I)V
ordinal=000 plt=0x00000000 bind=NONE type=FUNC name=Ljava/lang/Integer.method.toString()Ljava/lang/String;
ordinal=000 plt=0x000000d1 bind=NONE type=FUNC name=Ljava/lang/Object.method.<init>()V
ordinal=000 plt=0x000000d9 bind=NONE type=FUNC name=Ljava/lang/StringBuilder.method.<init>()V
ordinal=000 plt=0x00000000 bind=NONE type=FUNC name=Ljava/lang/StringBuilder.method.append(Ljava/lang/String;)Ljava/lang/StringBuilder;
ordinal=000 plt=0x00000000 bind=NONE type=FUNC name=Ljava/lang/StringBuilder.method.toString()Ljava/lang/String;

7 imports
[Symbols]
vaddr=0x00000168 paddr=0x00000168 ord=000 fwd= sz=0 bind= type=STATIC name=LHello.sfield_localVar2:I
vaddr=0x00000160 paddr=0x00000160 ord=000 fwd= sz=0 bind= type=FIELD name=LHello.ifield_localVar:I
vaddr=0x00000238 paddr=0x00000238 ord=000 fwd= sz=20 bind=GLOBAL type=FUNC name=LHello.method.<init>()V
vaddr=0x0000025c paddr=0x0000025c ord=000 fwd= sz=34 bind=GLOBAL type=FUNC name=LHello.method.main([Ljava/lang/String;)V
vaddr=0x00000290 paddr=0x00000290 ord=000 fwd= sz=0 bind=GLOBAL type=METH name=LHello.method.foo(I)V
vaddr=0x00000170 paddr=0x00000170 ord=000 fwd= sz=0 bind= type=FIELD name=LWorld.ifield_worldVar:I
vaddr=0x00000178 paddr=0x00000178 ord=000 fwd= sz=0 bind= type=FIELD name=LWorld.ifield_worldVar2:I
vaddr=0x000002f4 paddr=0x000002f4 ord=000 fwd= sz=20 bind=GLOBAL type=FUNC name=LWorld.method.<init>()V
vaddr=0x00000318 paddr=0x00000318 ord=000 fwd= sz=0 bind=GLOBAL type=METH name=LWorld.method.foo(I)V
vaddr=0x000000b9 paddr=0x000000b9 ord=000 fwd= sz=0 bind=NONE type=FUNC name=imp.Ljava/io/PrintStream.method.println(Ljava/lang/String;)V
vaddr=0x000000c1 paddr=0x000000c1 ord=000 fwd= sz=0 bind=NONE type=FUNC name=imp.Ljava/lang/Integer.method.<init>(I)V
vaddr=0x000000c9 paddr=0x000000c9 ord=000 fwd= sz=0 bind=NONE type=FUNC name=imp.Ljava/lang/Integer.method.toString()Ljava/lang/String;
vaddr=0x000000d1 paddr=0x000000d1 ord=000 fwd= sz=0 bind=NONE type=FUNC name=imp.Ljava/lang/Object.method.<init>()V
vaddr=0x000000d9 paddr=0x000000d9 ord=000 fwd= sz=0 bind=NONE type=FUNC name=imp.Ljava/lang/StringBuilder.method.<init>()V
vaddr=0x000000e1 paddr=0x000000e1 ord=000 fwd= sz=0 bind=NONE type=FUNC name=imp.Ljava/lang/StringBuilder.method.append(Ljava/lang/String;)Ljava/lang/StringBuilder;
vaddr=0x000000e9 paddr=0x000000e9 ord=000 fwd= sz=0 bind=NONE type=FUNC name=imp.Ljava/lang/StringBuilder.method.toString()Ljava/lang/String;

16 symbols```

As shows the snippet the longest imports are not working, any idea @radare ?

Thanks

**Comment by Maijin** [comment]
Ordinal should be incremented no?

**Comment by radare** [comment]
Yep

File: libr/bin/p/bin_dex.c [hunk 1]
```diff
@@ -986,12 +986,20 @@ static int dex_loadcode(RBinFile *arch, RBinDexObj *bin) {
...
-				imp->name = r_str_newf ("imp.%s.method.%s%s", class_name, method_name, signature);
-				// TODO: filter more chars that can be trash
+				imp->name  = r_str_newf ("%s.method.%s%s", class_name, method_name, signature);
 				imp->type = r_str_const ("FUNC");
 				imp->bind = r_str_const ("NONE");
-				//imp->paddr /= imp->vaddr = 0; // UNKNOWN
 				r_list_append (bin->imports_list, imp);
+
+				RBinSymbol *sym = R_NEW0 (RBinSymbol);
+				sym->name = r_str_newf("imp.%s", imp->name);
```
Comment (radare) [inline libr/bin/p/bin_dex.c]: missing space before (

File: libr/bin/p/bin_dex.c [hunk 2]
```diff
@@ -986,12 +986,20 @@ static int dex_loadcode(RBinFile *arch, RBinDexObj *bin) {
...
+				sym->type = r_str_const ("FUNC");
+				sym->bind = r_str_const ("NONE");
+				sym->paddr = sym->vaddr = bin->b->base + bin->header.header_size 
+					+ bin->header.strings_size + bin->header.class_size + 
+					(sizeof (struct dex_method_t) * i) ;
+				r_list_append (bin->methods_list, sym);
+
```
Comment (radare) [inline libr/bin/p/bin_dex.c]: empty line

Comment (h4ng3r) [inline libr/bin/p/bin_dex.c]: fixed.

**Comment by h4ng3r** [comment]
Guys, I pushed a new fix for the "Invalid address" problem. Maybe is not the best solution, maybe you have a better idea (@radare, @oddcoder), but I pushed this to have a starting point. thx

File: libr/anal/fcn.c [hunk 3]
```diff
@@ -468,6 +468,16 @@ static int fcn_recurse(RAnal *anal, RAnalFunction *fcn, ut64 addr, ut8 *buf, ut6
 		return R_ANAL_RET_ERROR; // MUST BE TOO DEEP
 	}
 
+	RListIter *iter;
+	RAnalPlugin *plugin;
+	r_list_foreach (anal->plugins, iter, plugin) {
+		if (plugin->is_valid_offset) {
+			int r = plugin->is_valid_offset(anal, addr, 0);
+			if (r < 0) return r;
```
Comment (radare) [inline libr/anal/fcn.c]: missing braces, you can just if (plugin->is_valid_offset) as long as the return value is not used by anything else

**Comment by Maijin** [comment]
Since the problem was in non-code section, why didn't you reused the eval variable "anal.noncode" which is already set by default on false otherwise you forbid completly user to analyse code in non-code section which may be needed in some cases.

File: libr/anal/fcn.c [hunk 4]
```diff
@@ -468,6 +468,16 @@ static int fcn_recurse(RAnal *anal, RAnalFunction *fcn, ut64 addr, ut8 *buf, ut6
 		return R_ANAL_RET_ERROR; // MUST BE TOO DEEP
 	}
 
+	RListIter *iter;
+	RAnalPlugin *plugin;
+	r_list_foreach (anal->plugins, iter, plugin) {
```
Comment (radare) [inline libr/anal/fcn.c]: dont iterate over all anal plugins, use only the current one

File: libr/anal/fcn.c (hunk 3 above)
Comment (radare) [inline libr/anal/fcn.c]: imho this function should be returning a boolean, not a 0 or -x

**Comment by radare** [comment]
Thanks!

**Pull request merged**